### ISSUE TITLE
Allow np.argmax to output top K maximum values

### ISSUE THREAD

**Issue by regstrtn** [numpy/numpy#15128]
Currently np.argmax, np.argmin, np.max and np.min return the maximum and minimum values respectively. 
`numpy.argmax(a, axis=None, out=None)`
Often times we need to get the topKmax or topKmin values, mostly in the case of machine learning workloads where you need top K classes for a given input. 
Requesting support for these scenarios in numpy. This scenario could be supported by np.max (and other functions), or there could be a separate family of methods np.maxK, minK etc?

**Comment by eric-wieser** [comment]
Are you looking for argpartition?

**Comment by regstrtn** [comment]
Hi, I just looked at np.argpartition. This method returns the kth smallest element. I was looking for a method that will return Top K smallest/largest elements in sorted order. It is possible to do it in numpy using argsort. 

```
In [1]: import numpy as np

In [2]: arr = np.array([1, 3, 2, 4, 5])

In [3]: arr.argsort()[-3:][::-1]
Out[3]: array([4, 3, 1])
```

This stackoverflow question is asking for the same thing. 
[https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array](https://stackoverflow.com/questions/6910641/how-do-i-get-indices-of-n-maximum-values-in-a-numpy-array)

**Comment by WarrenWeckesser** [comment]
[This answer](https://stackoverflow.com/a/23734295/1217358) to the stackoverflow question that you linked explains how to use the get the top K values using `argpartition`.  It also shows how to sort the top K values once you've found them.  Does that not answer your question?

**Comment by regstrtn** [comment]
Hi, both the argsort() and the argpartition() methods allow me to extract top K elements. 
My request was that since this is a common scenario, could we support this using a single method call? User simply calls np.argmaxK(arr, K) & np.maxK(arr, K) and gets the K largest elements in sorted order.

**Comment by mproszewska** [comment]
`argpartition` can also be used instead of current `argmax`, so I don't think that existence of `argpartition` is an argument against adding proposed parameter to `argmax`. The question is, whether this is needed addition. If so, I can create pull request.

**Comment by rossbar** [comment]
It seems like the desired behavior is achievable with existing tools. The next step in trying to push this forward would be to take the feature proposal to the mailing list (with a summary of the proposal, alternatives, and a link to this issue).

**Comment by andrewjong** [comment]
I support top k as an argument for readability. While the argpartition method works, it's rather convoluted.

**Comment by My3bIKAHT** [comment]
Returning indices of K min or max elements would be great addition

**Comment by akorman128** [comment]
Hi, I'd like to take this issue–– will create a branch and PR as per the contrib docs.

**Comment by rkern** [comment]
Please see gh-19117 and the [associated mailing list discussion](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/F4P5UVTAKRJJ3OORI6UOWFSUEE5CNTSC/) before starting any work.

**Comment by akorman128** [comment]
Okay so from what I've read in the mailing list there is consensus around creating 2 entirely new function: top_k(...) and argtop_k(...), which return the largest k values in order and their corresponding indices, respectively. 

With respect to @quarrying 's work in [https://github.com/numpy/numpy/pull/19117](url), it seems like the last comment by  @eric-wieser suggested breaking top_k into the aforementioned functions. 

Am I correct? If so I'll proceed with the edits.

**Comment by JuliaPoo** [comment]
I'm linking the discussion on array-api for `top_k` and related functions at data-apis/array-api#722, and the discussions on the [mailing list](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/F4P5UVTAKRJJ3OORI6UOWFSUEE5CNTSC/).

It seems that the consensus for such a function could look something like this:
```py
def top_k(
    x: array,
    k: int,
    /,
    *,
    axis: Optional[int] = None,
    mode: Literal["largest", "smallest"] = "largest",
) -> Tuple[array, array]
```

I don't mind drafting an implementation for this if it helps further the discussion. Some notes about `top_k` specifically for numpy.

I think `top_k` implementation can simply call `argpartition` internally, and we can let that decide a bunch of details about `top_k`. With reference to the discussion on [array-api]((https://github.com/data-apis/array-api/pull/722):

> Should we be more strict in specifying what should happen when `k` exceeds the number of elements?

`argpartition` does raise a `ValueError` when that happens.

> Are we okay with `None` being the default for axis, where the default behavior is searching over a flattened array?

Currently the default for `argparitition` is `axis=-1`, which performs the partitioning along the last axis rather than flattening. I think `top_k` for now can have a default `axis=-1` similarly until consensus is reached.

> Something that seems missing from this spec is to specify the handling of NaN values.

From experimenting with `np.max` and `np.min` it seems numpy returns false for any comparison with `nan`. Since `argpartition` uses this comparison as well it doesn't seem like `argpartition` is particularly intentional in its treatment of `nan`.

In addition to the current API, following [torch.topk](https://pytorch.org/docs/stable/generated/torch.topk.html) and [tf.math.top_k](https://www.tensorflow.org/api_docs/python/tf/math/top_k), we can consider the `out` keyword to specify an output array, and the `sorted` keyword. Putting it together can look something like:

```py
def top_k(
    x: array,
    k: int,
    /,
    *,
    axis: Optional[int] = -1,
    mode: Literal["largest", "smallest"] = "largest",
    sorted: bool = True,
    out: Optional[array] = None
) -> Tuple[array, array]
```

**Comment by mattip** [comment]
> we can consider the out keyword to specify an output array, ...

Given that the new function returns two arrays, I think this would have to be (like pytorch) an optional tuple of two arrays. Note neither `partition` nor `argpartition` take a `out` kwarg (they are part of the "sorting" family of functions, not ufuncs), so it might be unwieldy to try to implement an `out` kwarg.

**Comment by rgommers** [comment]
Thanks @JuliaPoo,  having a concrete implementation to try out will be helpful here. There may also be something to reuse from gh-19117, e.g. the `__array_function__` dispatcher.

**Comment by rgommers** [comment]
What would also be quite useful is to add tests for the new function in https://github.com/data-apis/array-api-tests/, which can easily (even in CI) be run against PyTorch, JAX, etc. so that it's easy to verify which semantics (e.g., treatment of `nan` values), or defaulting to largest/smallest, are consistent across libraries and which ones aren't.

**Comment by seberg** [comment]
Can we please clarify the NaN policy or at least voice opinions on it before adding this API?  IMO there are two consistent choices:
* Go with sorting, and push NaNs to the end, so that NaNs are not included in the result.  That also gives the same result as `nanmax`/`nanmin` with `N=1`.
* Go with min/max and always include NaNs, so that the function with `N=1` gives the same result as `max`/`argmax`.

Unfortunately, I really can't say I like the third choice as considering NaNs the largest element, which `top_k` does in any of the suggested implementations.  Conceptually, I cannot think of any logic by which that would make sense.
And sure, one can purposefully decide it is undefined, although users will rely on it whether we like it or not.

The current choices want to go with sorting (`nanmax`), but we don't define what _descending_ sorts should do (i.e. finding maximum values first):
* `descending_sort(x) == ascending_sort(x)[::-1]` except this fails for stable sorts.  This would give the conceptually inconsistent behavior that NaNs are not pushed to the end for being unordered but rather defined as the "largest value".
* `-descending_sort(-x) == ascending_sort(x)` (pushes NaNs to the end always; not actually a valid implementation for integers due to `-x` behavior)

(Yes, I understand that pushing NaNs to the end is annoying if you want to implement "top-k", because partition has no descending sort right now.)

**Comment by JuliaPoo** [comment]
I'm personally fine with considering `np.nan` as the largest element, mostly because it follows the sort order in other numpy apis like `np.sort/np.argsort` and `np.partition/np.argpartition`, and users might reasonably expect `top_k` to match the output of `sort`.

**Comment by seberg** [comment]
To add another point: The way I am suggesting to define it is also the way that dataframe libraries define `top_k` (e.g. `nsmallest` and `nlargest` in pandas) and sort order.
And those library authors care about NaN (or NULL value) semantics.

**Comment by JuliaPoo** [comment]
Would that look like adding a kwarg that's similar to `na_position` in `pd.sort_values`? Or maybe something like `ignore_nan:bool`? Specifying the `np.nan` position/ignoring them will probably take slightly longer while than the current implementation but shouldn't affect it too much.

**Comment by seberg** [comment]
We discussed this today, and I think the leaning was that the pandas behavior is the useful one:  It would be best to sort NaNs to the end (i.e. omit if possible) for both largest and smallest.
I don't know how simple that is in the sorting code, although, you can probably work around in python semi-reasonably (by special casing types that can contain NaNs).

**Comment by JuliaPoo** [comment]
If we're special casing a bunch of types a possible implementation can be:
```py
def top_k(a, k, /, *, axis=-1, largest=True):
    if k <= 0:
        raise ValueError(f'k(={k}) provided must be positive.')

    _arr = np.asanyarray(a)

    nan_containing = [
        np.clongdouble, np.complex128, np.complex64,
        np.float16, np.float32, np.float64, np.longdouble]
    to_negate = largest and _arr.dtype in nan_containing
    if to_negate:
        topk_values, topk_indices = top_k(-_arr, k, axis=axis, largest=False)
        return -topk_values, topk_indices

    # Actual implementation
    ...
```

However, this will ignore a bunch of situations like `np.array([np.uint8(1), np.nan], dtype=object)` where `np.nan` is again treated as the biggest element. I personally don't see any way one could come across these special situations though.
